firestore/bulkwriter.go

```diff
@@ -0,0 +1,319 @@
+// Note: You cannot write to (Create, Update, Set, or Delete) the same document more than once.
+func (bw *BulkWriter) Set(doc *DocumentRef, datum interface{}, opts ...SetOption) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newSetWrites(datum, opts)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot set %v on doc %v", datum, doc.ID)
+	}
+
+	if len(w) > 1 {
+		return nil, fmt.Errorf("firestore: too many writes sent to bulkwriter")
+	}
+
+	j := bw.write(w[0])
+	return j, nil
+}
+
+// Update adds a document update write to the queue of writes to send.
+// Note: You cannot write to (Create, Update, Set, or Delete) the same document more than once.
+func (bw *BulkWriter) Update(doc *DocumentRef, updates []Update, preconds ...Precondition) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newUpdatePathWrites(updates, preconds)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot update doc %v", doc.ID)
+	}
+
+	if len(w) > 1 {
+		return nil, fmt.Errorf("firestore: too many writes sent to bulkwriter")
+	}
+
+	j := bw.write(w[0])
+	return j, nil
+}
+
+// checkConditions determines whether this write attempt is valid. It returns
+// an error if either the BulkWriter has already been closed or if it
+// receives a nil document reference.
+func (bw *BulkWriter) checkWriteConditions(doc *DocumentRef) error {
+	if !bw.IsOpen() {
+		return errors.New("firestore: BulkWriter has been closed")
+	}
+	if doc == nil {
+		return errors.New("firestore: nil document contents")
+	}
+
+	_, havePath := bw.docUpdatePaths[doc.shortPath]
+	if havePath {
+		return fmt.Errorf("firestore: bulkwriter: received duplicate write for path: %v", doc.shortPath)
+	}
+
+	bw.docUpdatePaths[doc.shortPath] = true
+
+	return nil
+}
+
+// write packages up write requests into bulkWriterJob objects.
+func (bw *BulkWriter) write(w *pb.Write) *BulkWriterJob {
+
+	j := BulkWriterJob{
+		result: make(chan *pb.WriteResult, 1),
+		write:  w,
+		err:    make(chan error, 1),
+		ctx:    bw.ctx,
+	}
+
+	err := bw.bundler.Add(j, 0) // ignore operation size constraints; can't be inferred at compile time
+	if err != nil {
+		j.setError(err)
+	}
+	return &j
+}
+
+// send transmits writes to the service and matches response results to job channels.
+func (bw *BulkWriter) send(i interface{}) {
+	bwj := i.([]BulkWriterJob)
+
+	if len(bwj) == 0 {
+		return
+	}
+
+	var ws []*pb.Write
+	for _, w := range bwj {
+		ws = append(ws, w.write)
+	}
+
+	bwr := &pb.BatchWriteRequest{
+		Database: bw.database,
+		Writes:   ws,
+		Labels:   map[string]string{},
+	}
+
+	select {
+	case <-bw.ctx.Done():
+		return
+	default:
+		resp, err := bw.vc.BatchWrite(bw.ctx, bwr)
+		if err != nil {
+			// Do we need to be selective about what kind of errors we send?
+			for _, j := range bwj {
+				j.setError(err)
```

Comment:
What if all jobs error out. The buffer is set to 1 and will deadlock, right?

Reply:
I'm not quite sure what you mean by "buffer": the buffer of the channel being written to in `setError()`?

We will only follow this code path if the request itself is not successful. Each job would remain unresolved (because we won't have a response back from the server).

Reply:
Put another way: `BatchWrite()` requests have a one-to-many relationship with BulkWriterJobs.